core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -104,10 +110,13 @@
   // TODO(gpang): remove this set, and just use the indexes.
   private final ConcurrentHashSet<T> mObjects = new ConcurrentHashSet<>(8, 0.95f, 8);
   /**
-   * Map from {@link FieldIndex} to the index. An index is a map from index value to set of
-   * objects with that index value.
+   * Map from {@link FieldIndex} to the index. An index is a map from index value to one or a set of
+   * objects with that index value. A unique index is an index where each index value only maps to
+   * one object. A non-unique index is an index where an index value can map to one or more objects.
    */
-  private final Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> mIndexMap;
+  private final Map<UniqueFieldIndex<T>, ConcurrentHashMap<Object, T>> mIndexMapUnique;
```

Comment:
Are we trying to get this into 1.1.1? Doing something in a separate PR often means never doing it, and the code in this class is going to be quite messy until it's done :(